clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -664,6 +664,10 @@ void CodeGenFunction::EmitKernelMetadata(const FunctionDecl *FD,
     AttrMDArgs.push_back(
         llvm::ConstantAsMetadata::get(Builder.getInt(*XDimVal)));
 
+    for (auto i = AttrMDArgs.size(); i < 3; ++i)
+      AttrMDArgs.push_back(
+          llvm::ConstantAsMetadata::get(Builder.getInt(llvm::APInt(32, 1))));
```

Comment:
I always get confused with this: above we add, 'z', then 'y' and then 'x'.  But if 'z' or 'y' were missing, when you fill '1' you are adding the 1s after 'x'.   Does this lead to the right behavior?

Reply:
If it's any consolation, I have to refresh my understanding of this every thing it pops up. The dimension-flipping happens based on the dimensionality, so:
3D: x -> z, y -> y and z -> x, e.g. {32, 16, 8} -> {8, 16, 32}. No padding needed.
2D: x -> y and y -> x, e.g. {32, 16} -> {16, 32}. Padding is added at the end, i.e. {16, 32, 1}.
1D: x -> x, e.g. {32} -> {32}. Padding is added at the end, i.e. {32, 1, 1}.

Reply:
Yeap, it's exactly how @steffenlarsen explains it.

The way to think about it is that SYCL allows for non-3-dims wg sizes, however we are using openCL's mechanism to pass it to the backend. So, the flipping has to happen first in order to preserve the SYCL's semantics, only after that we pad, to align with openCL.